README.md
# fitplotter

A simple one-page tool to plot parameters recorded in the `*.fit` file used by Garmin and Suunto.

Starting from 2021-03-03, it also works with `*.tkl` format (watch models: GPSmaster, navrun-500, ultrasport, etc).

Version [online](https://karaul.github.io/fitplotter). If you have no own fit files at the moment, you can download them from [examples](https://github.com/karaul/fitplotter/blob/main/examples/).

From 2021-03-08 you can download `*.fit` file from your _Garmin Connect_ account, for how to do it read  __Usage__ section below.

Check the wrapper for `fitplotter` is [fitalyser](https://github.com/karaul/fitalyser) - parallel project.

Check  [GChandler](https://github.com/karaul/gchandler) - download automatically new activities from Garnmin Connect to the local storage


![Alt text](https://github.com/karaul/fitplotter/blob/main/screenshots/screenshot.JPG?raw=true)

## Motivation

Garmin device does  recording the measured paremeters into a binary `.fit` file as a function of time and distance. Then, in Garmin Connect you can visualise these metrics in graphs. I found Garmin Connect a bit inconvenient for everyday analysis. For instance, I could not check heart rate efficiency (HRE) (if you do not know read about HRE [here](https://drive.google.com/file/d/17wK0y5p7rYlMRBogpZ9yicdnV191E1qx/view?usp=sharing) and  [in Russian here](https://grumbler.livejournal.com/104934.html)) during the run. Then, the data from different days could not be visualised in one plot. Also, it is not convenient in Garmin Connect to see simultaneously a pointer in the plot and on the GPS track. Moreover, Garmin Connect is slow, and it's always better to keep your files in your local storage and have opportunity to check them.  And finally, I wanted to learn JS programming.

## Features

- Work with any `*.fit` file (Garmin, Suunto);
- Work with `*.tkl` (sport watch models: GPSmaster, navrun-500, ultrasport, etc);
- Plot data from different files in one plot;
- Independent y-axes;
- hide/unhide curves by clicking on curve's name in the below legend;
- Synchronization of the plot pointer at the left with the map marker on the right;
- zoom of plot area by mouse click;
- Automatically calculate an average of the plotted parameters in the zoomed area;
- Visualise laps;
- Can remove spikes and  make noisy data smooth;
- Calculate breath_rate during the run, if HRV is on in the settings;

## Installation

### Installation for non-programmers who found this program in github

- click on the green button and download everything as zip file
- unzip
- Double click on index.html

### Installation for programmers who found it in github

`github clone https://github.com/karaul/fitplotter.git`

## Usage

Press button "Choose file" and select a `*.fit` file to analyse. In the Garmin unit,  `*.fit` files are in the folder "activities", which can be seen when you connect Garmin  with PC via USB. Or you can download your fit files from Garmin Connect. Also, you can play with `*.fit` files located for the demonstration purpose in the folder [examples](https://github.com/karaul/fitplotter/blob/main/examples/)

If everything works fine, after you open the `.fit` file, you will see a graph on the left and map on the right. Few screenshots are in the folder [screenshots](https://github.com/karaul/fitplotter/blob/main/screenshots/). Play with droplists: x-axis, y-axis, automode/manual add. Zoom area in the left graph and then see the averaged over the zoomed area on the right map. Click on the legend with different choice in the bottom droplist.

### In order to download `*.fit` file from _Garmin Connect_ account

- Login to _Garmin Connect_ (GC) account;
- Navigate to the activity URL `https://connect.garmin.com/modern/activity/xxxxxxxxxx`;
- The latest 10-digits number `xxxxxxxxx`in the _Garmin Connect_ URL is `acitvity ID`;
- Type in the `fitplotter`'s URL field the noticed `activity ID`;
- Click Download, wait a bit and save your zipped activity taken from _Garmin Connect_ into the local folder;
- Unzip the activity in your local folder;
- Click "Choose file" and load the activity as a local file into `fitplotter`

Russians may also check [ЧАВО](https://github.com/karaul/fitplotter/blob/main/screenshots/FAQ_ru.md)

## Contributors

[fit-file-parser](https://github.com/jimmykane/fit-parser) by Dimitrios Kanellopoulos. This  was taken and modified to get it working with HTML5 but with http static server.

[BackFitClientSide](https://github.com/gfmoore/BackFitClientSide) by Gordon Moore who demonstrated how to work without a server

[Sarah Lucke](https://github.com/SarahLucke) who explained me how to make an online fitplotter version based on the github repository

Thanks are to CanvasJS and Leaflet libraries for smooth work of the application.

## License

MIT license

(c) Evgeny Votyakov aka karaul ([about me](http://www.irc-club.ru/karaul.html) in Russian), Nicosia, Cyprus, 2021


CHANGELOG.md
# Changelog

Changes of the `fitplotter` project are documented in this file

## To do

- οn-fly unzip of the downloaded file
- ~~immediate analysis of the downloaded file~~ (impossible without an own server)

## 2021-03-24

- For x-axis = timestamp, now there appears breath_rate in y-axis droplist

## 2021-03-22

- Added RR intervals and respiratory rate from HRV data. Select the corresponding x-axis

## 2021-03-11

- strava download  is implemented for  a single activity

## 2021-03-07

- It is implemented download of a single activity from Garmin Connect

## 2021-03-03

- tkl-parser is added;
- massive refactoring to split loadfile function from the calculations;
- added "update Map" button;
- fixed bug with loaded files droplist, now each loaded file  brings the own yaxes droplist. This enables to compare data on the same tracks but with different watches, what is especially useful by working with the data in the span of several years.  

## 2021-03-01

- Added droplist for the loaded files;
- Added button to clean map;
- CHANGELOG started;

## 2021-02-12

Project transferred to github


src/bf-fit-parser.js
//hacked by Gordon Moore 18 Feb 2021 to allow client side
//in html just <script src=./[your scripts folder/bf-fit-parser.js></script>

//I made a change to code at line 22 as I already have obtained an array buffer - using async await code 
//I don't know how to refactor to make it work without the change.

//import { getArrayBuffer, calculateCRC, readRecord } from './binary';

//export default 
class FitParser {
  constructor(options = {}) {
    this.options = {
      force: options.force != null ? options.force : true,
      speedUnit: options.speedUnit || 'm/s',
      lengthUnit: options.lengthUnit || 'm',
      temperatureUnit: options.temperatureUnit || 'celsius',
      elapsedRecordField: options.elapsedRecordField || false,
      mode: options.mode || 'list',
    };
  }

  parse(content, callback) {
    //const blob = new Uint8Array(getArrayBuffer(content));  //content is already an arraybuffer in my code!
    const blob = new Uint8Array(content);

    if (blob.length < 12) {
      callback('File too small to be a FIT file', {});
      if (!this.options.force) {
        return;
      }
    }

    const headerLength = blob[0];
    if (headerLength !== 14 && headerLength !== 12) {
      callback('Incorrect header size', {});
      if (!this.options.force) {
        return;
      }
    }

    let fileTypeString = '';
    for (let i = 8; i < 12; i++) {
      fileTypeString += String.fromCharCode(blob[i]);
    }
    if (fileTypeString !== '.FIT') {
      callback('Missing \'.FIT\' in header', {});
      if (!this.options.force) {
        return;
      }
    }

    if (headerLength === 14) {
      const crcHeader = blob[12] + (blob[13] << 8);
      const crcHeaderCalc = calculateCRC(blob, 0, 12);
      if (crcHeader !== crcHeaderCalc) {
        // callback('Header CRC mismatch', {});
        // TODO: fix Header CRC check
        if (!this.options.force) {
          return;
        }
      }
    }

    const protocolVersion = blob[1];
    const profileVersion = blob[2] + (blob[3] << 8);
    const dataLength = blob[4] + (blob[5] << 8) + (blob[6] << 16) + (blob[7] << 24);
    const crcStart = dataLength + headerLength;
    const crcFile = blob[crcStart] + (blob[crcStart + 1] << 8);
    const crcFileCalc = calculateCRC(blob, headerLength === 12 ? 0 : headerLength, crcStart);

    if (crcFile !== crcFileCalc) {
      // callback('File CRC mismatch', {});
      // TODO: fix File CRC check
      if (!this.options.force) {
        return;
      }
    }

    const fitObj = {};
    fitObj.protocolVersion = protocolVersion;
    fitObj.profileVersion = profileVersion;

    const sessions = [];
    const laps = [];
    const records = [];
    const events = [];
    const hrv = [];
    const devices = [];
    const applications = [];
    const fieldDescriptions = [];
    const dive_gases = [];
    const course_points = [];
    const sports = [];
    const monitors = [];
    const stress = [];
    const definitions = [];
    const file_ids = [];
    const monitor_info = [];
    const lengths = [];

    let tempLaps = [];
    let tempLengths = [];
    let tempRecords = [];

    let loopIndex = headerLength;
    const messageTypes = [];
    const developerFields = [];

    const isModeCascade = this.options.mode === 'cascade';
    const isCascadeNeeded = isModeCascade || this.options.mode === 'both';

    let startDate, lastStopTimestamp;
    let pausedTime = 0;

    while (loopIndex < crcStart) {
      const { nextIndex,
        messageType,
        message } = readRecord(blob, messageTypes, developerFields, loopIndex, this.options, startDate, pausedTime);
      loopIndex = nextIndex;

      switch (messageType) {
        case 'lap':
          if (isCascadeNeeded) {
            message.records = tempRecords;
            tempRecords = [];
            tempLaps.push(message);
            message.lengths = tempLengths;
            tempLengths = [];
          }
          laps.push(message);
          break;
        case 'session':
          if (isCascadeNeeded) {
            message.laps = tempLaps;
            tempLaps = [];
          }
          sessions.push(message);
          break;
        case 'event':
          if (message.event === 'timer') {
            if (message.event_type === 'stop_all') {
              lastStopTimestamp = message.timestamp;
            } else if (message.event_type === 'start' && lastStopTimestamp) {
              pausedTime += (message.timestamp - lastStopTimestamp) / 1000;
            }
          }
          events.push(message);
          break;
        case 'length':
          if (isCascadeNeeded) {
            tempLengths.push(message);
          }
          lengths.push(message);
          break;
        case 'hrv':
          hrv.push(message);
          break;
        case 'record':
          if (!startDate) {
            startDate = message.timestamp;
            message.elapsed_time = 0;
            message.timer_time = 0;
          }
          records.push(message);
          if (isCascadeNeeded) {
            tempRecords.push(message);
          }
          break;
        case 'field_description':
          fieldDescriptions.push(message);
          break;
        case 'device_info':
          devices.push(message);
          break;
        case 'developer_data_id':
          applications.push(message);
          break;
        case 'dive_gas':
          dive_gases.push(message);
          break;
        case 'course_point':
          course_points.push(message);
          break;
        case 'sport':
          sports.push(message);
          break;
        case 'file_id':
          if(message){
            file_ids.push(message);
          }
          break;
        case 'definition':
          if(message){
            definitions.push(message);
          }
          break;
        case 'monitoring':
          monitors.push(message);
          break;
        case 'monitoring_info':
          monitor_info.push(message);
          break;
        case 'stress_level':
          stress.push(message);
          break;
        case 'software':
          fitObj.software = message;
          break;
        default:
          if (messageType !== '') {
            fitObj[messageType] = message;
          }
          break;
      }
    }

    if (isCascadeNeeded) {
      fitObj.activity = fitObj.activity || {};
      fitObj.activity.sessions = sessions;
      fitObj.activity.events = events;
      fitObj.activity.hrv = hrv;
      fitObj.activity.device_infos = devices;
      fitObj.activity.developer_data_ids = applications;
      fitObj.activity.field_descriptions = fieldDescriptions;
      fitObj.activity.sports = sports;
    }
    if (!isModeCascade) {
      fitObj.sessions = sessions;
      fitObj.laps = laps;
      fitObj.lengths = lengths;
      fitObj.records = records;
      fitObj.events = events;
      fitObj.device_infos = devices;
      fitObj.developer_data_ids = applications;
      fitObj.field_descriptions = fieldDescriptions;
      fitObj.hrv = hrv;
      fitObj.dive_gases = dive_gases;
      fitObj.course_points = course_points;
      fitObj.sports = sports;
      fitObj.devices = devices;
      fitObj.monitors = monitors;
      fitObj.stress = stress;
      fitObj.file_ids = file_ids;
      fitObj.monitor_info = monitor_info;
      fitObj.definitions = definitions;
    }

    callback(null, fitObj);
  }
}
